add event source on Lambda examples Gen 2 (#7184)
aws-amplify/docs: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx

In this guide, lets configure a Lambda function with an Amazon Kinesis stream as an event source. The Lambda function is automatically triggered whenever new data is published to the stream, whether you're processing streaming data, reacting to application events, or automating workflows.

To get started, install the AWS Lambda Powertools Logger package, which provides a structured logging capabilities for your Lambda function.

```bash title="Terminal" showLineNumbers={false}
npm add @aws-lambda-powertools/logger
```

Second, create a new directory and a resource file, `amplify/functions/kinesis-function/resource.ts`. Then, define the Function with `defineFunction`:

```ts title="amplify/functions/kinesis-function/resource.ts"
import { defineFunction } from "@aws-amplify/backend";

export const myKinesisFunction = defineFunction({
  name: "kinesis-function",

>   entry: "./handler.ts",

Reviewer: ```suggestion
```

});
```

Third, create the corresponding handler file, `amplify/functions/kinesis-function/handler.ts`, file with the following contents:

```ts title="amplify/functions/kinesis-function/handler.ts"

> import {
>   KinesisStreamEvent,
>   Context,
>   KinesisStreamHandler,
>   KinesisStreamRecordPayload,
>   KinesisStreamBatchResponse,
> } from "aws-lambda";
> import { Buffer } from "buffer";
> import { Logger } from "@aws-lambda-powertools/logger";
> 
> const logger = new Logger({
>   logLevel: "INFO",
>   serviceName: "kinesis-stream-handler",
> });
> 
> export const handler: KinesisStreamHandler = async (
>   event: KinesisStreamEvent,
>   context: Context
> ): Promise<KinesisStreamBatchResponse> => {
>   for (const record of event.Records) {
>     try {
>       logger.info(`Processed Kinesis Event - EventID: ${record.eventID}`);
>       const recordData = await getRecordDataAsync(record.kinesis);
>       logger.info(`Record Data: ${recordData}`);
>     } catch (err) {
>       logger.error(`An error occurred ${err}`);
>       /*
>       When processing stream data, if any item fails, returning the failed item's position immediately
>       prompts Lambda to retry from this item forward, ensuring continuous processing without skipping data.
>       */
>       return {
>         batchItemFailures: [{ itemIdentifier: record.kinesis.sequenceNumber }],
>       };
>     }
>   }
>   logger.info(`Successfully processed ${event.Records.length} records.`);
>   return { batchItemFailures: [] };
> };
> 
> async function getRecordDataAsync(
>   payload: KinesisStreamRecordPayload
> ): Promise<string> {
>   var data = Buffer.from(payload.data, "base64").toString("utf-8");
>   await Promise.resolve(1); //Placeholder for an async process
>   return data;
> }

Reviewer: ```suggestion
import type {
  KinesisStreamBatchResponse,
  KinesisStreamHandler,
  KinesisStreamRecordPayload,
} from "aws-lambda";
import { Buffer } from "node:buffer";
import { Logger } from "@aws-lambda-powertools/logger";

const logger = new Logger({
  logLevel: "INFO",
  serviceName: "kinesis-stream-handler",
});

export const handler: KinesisStreamHandler = async (
  event,
  context
): Promise<KinesisStreamBatchResponse> => {
  for (const record of event.Records) {
    try {
      logger.info(`Processed Kinesis Event - EventID: ${record.eventID}`);
      const recordData = await getRecordDataAsync(record.kinesis);
      logger.info(`Record Data: ${recordData}`);
    } catch (err) {
      logger.error(`An error occurred ${err}`);
      /**
			 * When processing stream data, if any item fails, returning the failed item's position immediately
			 * prompts Lambda to retry from this item forward, ensuring continuous processing without skipping data.
			 */
      return {
        batchItemFailures: [{ itemIdentifier: record.kinesis.sequenceNumber }],
      };
    }
  }
  logger.info(`Successfully processed ${event.Records.length} records.`);
  return { batchItemFailures: [] };
};

async function getRecordDataAsync(
  payload: KinesisStreamRecordPayload
): Promise<string> {
  const data = Buffer.from(payload.data, "base64").toString("utf-8");
  await Promise.resolve(1); // Placeholder for an async process
  return data;
}
```

```

Lastly, create the Kinesis stream in the `amplify/backend.ts` file:

```ts title="amplify/backend.ts"

> import { defineBackend } from "@aws-amplify/backend";
> import { auth } from "./auth/resource";
> import { data } from "./data/resource";
> import { myKinesisFunction } from "./functions/kinesis-function/resource";
> import { KinesisEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
> import { StartingPosition } from "aws-cdk-lib/aws-lambda";
> 
> const backend = defineBackend({
>   auth,
>   data,
>   myKinesisFunction,
>   //other resources
> });

Reviewer: ```suggestion
import { defineBackend } from "@aws-amplify/backend";
import { KinesisEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { StartingPosition } from "aws-cdk-lib/aws-lambda";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import { myKinesisFunction } from "./functions/kinesis-function/resource";

const backend = defineBackend({
  auth,
  data,
  myKinesisFunction,
});
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {

>   title: 'Lambda event source',

Reviewer: can we change this title to be a bit more descriptive? maybe just "Event handlers"?

## DynamoDB stream

You can also configure a Lambda function with an Amazon DynamoDB stream as an event source. The Lambda function is automatically triggered whenever an item is added, updated, or deleted from the table, enabling you to build real-time applications that react to changes in your data. In this example, we will use a `Todo` table created by a data model on the GraphQL API.

To get started, install the AWS Lambda Powertools Logger package, which provides a structured logging capabilities for your Lambda function.

```bash title="Terminal" showLineNumbers={false}
npm add @aws-lambda-powertools/logger
```

Second, create a new directory and a resource file, `amplify/functions/dynamoDB-function/resource.ts`. Then, define the Function with `defineFunction`:

```ts title="amplify/functions/kinesis-function/resource.ts"
import { defineFunction } from "@aws-amplify/backend";

export const myDynamoDBFunction = defineFunction({
  name: "dynamoDB-function",
});
```

Third, create the corresponding handler file, `amplify/functions/dynamoDB-function/handler.ts`, file with the following contents:

```ts title="amplify/functions/dynamoDB-function/resource.ts"

> import { Logger } from "@aws-lambda-powertools/logger";
> import { DynamoDBStreamHandler } from "aws-lambda";

Reviewer: ```suggestion
import type { DynamoDBStreamHandler } from "aws-lambda";
import { Logger } from "@aws-lambda-powertools/logger";
```

explicit type import nit

const logger = new Logger({
  logLevel: "INFO",
  serviceName: "dynamodb-stream-handler",
});

export const handler: DynamoDBStreamHandler = async (event) => {

>   const batchItemFailures: { itemIdentifier: string }[] = [];

Reviewer: is this needed?
